MechJeb2/VesselState.cs

```diff
@@ -913,6 +913,32 @@ public void RCSTorque()
             GUILayout.EndVertical();
         }
 
+        //FARCalculateVesselAeroForces(vessel,out farForce,out farTorque,surfaceVelocity,altitudeASL);
+        private Vector3 lastFarForce, lastFarTorque;
+        private Vector3d lastSurfaceVelocity;
+        private double lastAltitudeASL;
+        private const double _dVelocitySqrThreshold = 100;
+        private const double _dVelocitySqrMinThreshold = 0.0001;
+        private const double _dAltitudeThreshold = 300;
+        private const float _fOrientationThreshold = 5;
+        private void CalculateVesselAeroForcesWithCache(Vessel v, out Vector3 farForce, out Vector3 farTorque, Vector3d surfaceVelocity, double altitudeASL) 
+        {
+            if ((lastSurfaceVelocity - surfaceVelocity).sqrMagnitude > _dVelocitySqrThreshold
+                || Math.Abs(altitudeASL - lastAltitudeASL) > _dAltitudeThreshold
+                || (surfaceVelocity.sqrMagnitude > _dVelocitySqrMinThreshold && Vector3.Angle(lastSurfaceVelocity, surfaceVelocity) > _fOrientationThreshold))
```

Comment:
what if your surface velocity is just small and bouncing around Vector3d.zero?  consider some sort of hovering vehicle  for example.

it sort of feels like it would be better to have a check for the angle that wasn't constrained, then a check for the change in the magnitude that was expressed more as a percentage (or a squared percentage maybe), and then the change in altitude check.

Reply:
I initially didn't constrain the angle check.  The realized low surface velocity hovering around 0 was going to trigger the angle check probably too often, so constrained it to a min surface velocity.  (_dVelocitySqrMinThreshold = 0.0001 is a 0.01m/s threshold.)

If you'd rather unconstrained and trigger the recalculate more often when truly not moving but only perfectly hovering, that's also fine.

Could also add a check and new parameter for the up direction on the root part, instead of assessing an orientation change from the change in velocity?  Would that be better?

Reply:
Thinking a bit more, these are only the aero forces, this isn't the entire computation.  Force and torque at near-zero velocity are zero, so you don't care.  You do, however, care about change of AoA at velocity great enough to have force/torque.

`(surfaceVelocity.sqrMagnitude > _dVelocitySqrMinThreshold && Vector3.Angle(vessel.rootPart.transform.TransformDirection(Vector3.up), surfaceVelocity) - lastAoA > _fAoAThreshold))`

That instead?  dVelocitySqrMinThreshold ~= 1.  fAoAThreshold (in degrees) ~= 2?